core/variant_parser.cpp

```diff
@@ -35,36 +35,70 @@
 #include "core/os/keyboard.h"
 #include "core/string_buffer.h"
 
-CharType VariantParser::StreamFile::get_char() {
-	return f->get_8();
+CharType VariantParser::Stream::get_char() {
+	// is within buffer?
+	if (readahead_pointer < readahead_filled) {
+		return readahead_buffer[readahead_pointer++];
+	}
+
+	// attempt to readahead
+	readahead_filled = _read_buffer(readahead_buffer, READAHEAD_SIZE);
+	if (readahead_filled) {
+		readahead_pointer = 0;
+	} else {
+		// EOF
+		readahead_pointer = 1;
+		eof = true;
+		return 0;
+	}
+	return get_char();
```

Comment:
I'm a bit surprised by the recursive call TBH :stuck_out_tongue:

Could be replaced by the following structure. But it's also fine to keep it the way it is.

```gdscript
if pointer >= filled:
    filled = _read_buffer()
    if filled == 0:
        pointer = 1
        eof = true
        return 0
    pointer = 0

return buffer[pointer++]
```

Reply:
Yeah, same result.

One of the reasons I got into the habit of putting the expected result to be the if statement (rather than the else) is that long ago I read the assembly generated tends to favour that branch (much like the `unlikely` keyword, before that was available). Whether that is still the case I don't know, have to test in godbolt! :grinning: